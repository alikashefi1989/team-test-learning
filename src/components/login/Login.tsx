import { useEffect, useRef, useState } from "react"

export default function Login() {
    const username = useRef<any>(null);
    const password = useRef<any>(null);
    const [loginData, setLoginData] = useState<any>(null);

    useEffect(() => {
        username.current?.focus();
    }, []);

    async function submitForm() {
        const data = {
            id: new Date().getTime(),
            username: username.current.value,
            password: password.current.value
        }

        try {
            const res = await fetch('http://localhost:5004/users', {
                method: 'POST',
                body: JSON.stringify(data),
                headers: {
                    'Content-Type': 'application/json'
                },
            });
            setLoginData(await res.json());
        } catch (error) {
        }

    }

    return <div>
        <div>
            <label htmlFor="username">username</label>
            <input type="text" name="username" id="username"
                ref={username}
                data-testid="username-input"
            />
        </div>
        <div>
            <label htmlFor="password">password</label>
            <input type="text" id="password" name="password"
                ref={password}
                data-testid="password-input"
            />
        </div>

        <div>
            <button type="submit" onClick={submitForm}
                data-testid="submit-button"
            >submit</button>
        </div>

        <br /><br />
        {loginData && <pre data-testid="data-response">{JSON.stringify(loginData)}</pre>}
    </div>
}
