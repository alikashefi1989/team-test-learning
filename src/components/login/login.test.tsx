import { render, fireEvent,waitFor } from '@testing-library/react';
import React from 'react';
import { create } from 'react-test-renderer';
import Login from './Login';

describe('login component', () => {
    const renderComponent = (props?: any) => create(<Login {...props} />);

    it('should render login component', () => {
        const component = renderComponent().toJSON();
        expect(component).toMatchSnapshot()
    });

    it('should find the login page with only two inputs', () => {
        const { getByTestId, container } = render(<Login />);
        const usernameInput = getByTestId('username-input');
        const passwordInput = getByTestId('password-input');

        expect(usernameInput).toBeInTheDocument();
        expect(passwordInput).toBeInTheDocument();

        const otherInputs = container.querySelectorAll('input');
        expect(otherInputs.length).toBe(2);
    });

    it('should find the inputs empty', () => {
        const { getByTestId } = render(<Login />);
        const usernameInput = getByTestId('username-input');
        const passwordInput = getByTestId('password-input');

        expect(usernameInput).toHaveValue('');
        expect(passwordInput).toHaveValue('');
    });

    it('should fill and submit the form', async() => {
        const { getByTestId } = render(<Login />);
        const usernameInput = getByTestId('username-input');
        const passwordInput = getByTestId('password-input');
        const submitButton = getByTestId('submit-button');

        fireEvent.change(usernameInput, { target: { value: 'abolhasan' } });
        fireEvent.change(passwordInput, { target: { value: '1357' } });

        fireEvent.click(submitButton);

        const dataResponse = await waitFor<HTMLElement>(() => getByTestId('data-response'));
        expect(dataResponse).toBeInTheDocument();
    });
});
