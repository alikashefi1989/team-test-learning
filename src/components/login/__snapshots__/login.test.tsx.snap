// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login component should render login component 1`] = `
<div>
  <div>
    <label
      htmlFor="username"
    >
      username
    </label>
    <input
      data-testid="username-input"
      id="username"
      name="username"
      type="text"
    />
  </div>
  <div>
    <label
      htmlFor="password"
    >
      password
    </label>
    <input
      data-testid="password-input"
      id="password"
      name="password"
      type="text"
    />
  </div>
  <div>
    <button
      data-testid="submit-button"
      onClick={[Function]}
      type="submit"
    >
      submit
    </button>
  </div>
  <br />
  <br />
</div>
`;
